src/EventListener/UserSessionListener.php

```diff
@@ -0,0 +1,274 @@
+                $refererOld[$refererId],
+                $refererOld[$ref]
+            );
+            $refererNew[$refererId]['last'] = $refererOld[$ref]['current'];
+        } elseif (count($refererOld) > 1) {
+            $refererNew[$refererId] = end($refererOld);
+        }
+
+        $refererNew[$refererId]['current'] = $request->getRequestUri();
+
+        $bag->set($key, $refererNew);
+
+        $this->storeSession();
+    }
+
+    /**
+     * Check if we should not modify the session in the back end.
+     *
+     * @param Request $request
+     *
+     * @return bool
+     */
+    private function checkShouldNotModifyBackendSession(Request $request)
+    {
+        if (!$request->query->has('act')
+            && !$request->query->has('key')
+            && !$request->query->has('token')
+            && !$request->query->has('state')
+            && 'feRedirect' !== $request->query->get('do')
+            && !$request->isXmlHttpRequest()
+        ) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Stores the session for the front end.
+     *
+     * @param Request $request
+     */
+    private function storeFrontendSession(Request $request)
+    {
+        /** @var AttributeBagInterface $bag */
+        $bag = $this->getSessionBag();
+
+        $refererOld = $bag->get('referer');
+
+        if (!$this->checkShouldNotModifyFrontendSession($request, $refererOld)) {
+            $this->storeSession();
+            return;
+        }
+
+        $refererNew = [
+            'last'      => $refererOld['current'],
+            'current'   => $request->getRequestUri()
+        ];
+
+        $this->session->set('referer', $refererNew);
+
+        $this->storeSession();
+    }
+
+    /**
+     * Check if we should not modify the session in the front end.
+     *
+     * @param Request $request
+     * @param array   $refererOld
+     *
+     * @return bool
+     */
+    private function checkShouldNotModifyFrontendSession(
+        Request $request,
+        $refererOld = null
+    ) {
+        if (null === $refererOld) {
+            return false;
+        }
+
+        if (!$request->query->has('pdf')
+            && !$request->query->has('file')
+            && !$request->query->has('id')
+            && $refererOld['current'] !==  $request->getRequestUri()
+        ) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Gets the session bag.
+     *
+     * @return AttributeBagInterface
+     */
+    private function getSessionBag()
+    {
+        $bag = 'contao_backend';
+
+        if ($this->isFrontendScope()) {
+            $bag = 'contao_frontend';
+        }
+
+        return $this->session->getBag($bag);
+    }
+
+    /**
+     * Stores the session data in the database.
+     */
+    private function storeSession()
+    {
+        $user   = $this->getUserObject();
+        $table  = $this->isFrontendScope() ? 'tl_member' : 'tl_user';
```

Comment:
The old code uses old classes such as `Database`, `Session` and `Input` and the constant  `TL_REFERER_ID`. I already did all the refactoring work and it's a lot better now and we can even write tests for it. Especially as synchronizing with the Contao 3 repository is not an argument at all because if it was one it would mean that we can never move on to new things because if we do syncing with Contao 3 is complicated :-)